[Feature] Additional zeus interface actions

Joecuronium opened the pull request:
### Pull Request Description
**When merged this pull request will:**
Add the CAFE Immersive revive to the medical context actions
Move ending config from the description.ext into a endings file in the configuration folder
Allow any number of endings, with custom music and code execution
Endings will be run from Zeus ACE-actions with a dialog to choose the ending
Add a module to Zeus that allow the toogling of AI pathing ability
Closes #66 

### Release Notes
Endings are now defined in configuration\endings.hpp
Ending triggers are removed from the map screen admin menu and added to Zeus ACE-interact
New module to toggle AI pathing ability
New medical context action for immersive revive

## IMPORTANT

- [x] The Release Notes section below must be filled out appropriately to explain the changes made in this PR. This section will be used in Framework Changelogs.
- [x] If the contribution affects [the wiki](https://github.com/CombinedArmsGaming/CAFE3/wiki), please include your changes in this pull request so the wiki is consistently updated.
- [x] [Contribution Guidelines](https://github.com/CombinedArmsGaming/CAFE3/blob/release/contributing.md) are read, understood and applied.
- [x] Title of this PR uses our standard template `[Descriptor] - Add|Fix|Improve|Change|Make|Remove {changes}`.

Joecuronium committed:
Added immersive Revive ZEN context option

Joecuronium committed:
Added "Toggle AI pathing ability" module

Joecuronium committed:
added ending Dialog

Joecuronium committed:
reworked Ending system and transferred into ZEUS actions

Joecuronium commented:
Wiki commit will be done at work today

Joecuronium committed:
changed var names to better descriptors

Joecuronium committed:
variable names fixed

Joecuronium commented:
Wiki done, please review

Joecuronium committed:
Why is my gitignore in here lol

Bubbus commented on components/miscClient/zen/fn_zen_toggleAIPath.sqf:
```diff
@@ -0,0 +1,28 @@
+    ["No valid unit selected"] call zen_common_fnc_showMessage;
+};
+
+private _alreadyDisabled = !(_unit checkAIFeature "PATH"); 
```
The docs don't say whether `_unit` can be a non-local argument to the `checkAIFeature` command - it always will be if the unit was not spawned via Zeus interface.  I have observed cases where the Zeus UI doesn't know the proper state of the AI features, so this may be a problem.

Consider moving this check into `f_fnc_toggleAIPath`, so you can avoid the question and also make it so `f_fnc_toggleAIPath` truly does toggle the AI pathing.

Bubbus commented on components/miscClient/zen_modules.sqf:
```diff
@@ -65,7 +74,29 @@ call
+
+    private _immersiveReviveCondition =
+    {
+        (_objects findIf {isPlayer _x} != -1);
```
This might be useful upon AI when AI-unconsciousness is enabled, this is a very niche use case though so I don't think this needs to be changed now - let's just make a ticket about it that we can pick up when we have nothing better to do.

Joecuronium replied:
Just heal the unit then, problem solved. Let's not overcomplicate things.

Bubbus commented on components/miscShared/fn_toggleAIPath.sqf:
```diff
@@ -0,0 +1,15 @@
+
+
+params ["_unit", "_alreadyDisabled"];
+RUN_LOCAL_TO(_unit,_this,f_fnc_disableAI);
```
Potential misnaming - this will send the `f_fnc_disableAI` function to `_unit`'s locality instead of `f_fnc_toggleAIPath`.

The arguments are out of order, they should be `RUN_LOCAL_TO(OBJ,FUNC,ARGS)`.  There was the same issue with ordering in some of the Zeus medical functions, you may have copied from there?

Joecuronium replied:
In renaming the functions 5 times, I missed this thanks. Also yes, it was copied from the medical.

Bubbus commented on components/miscShared/fn_toggleAIPath.sqf:
```diff
@@ -0,0 +1,15 @@
+RUN_LOCAL_TO(_unit,_this,f_fnc_disableAI);
+
+
+if (_alreadyDisabled) then
```
If this function is supposed to set the PATH state instead of toggling it, I would suggest the name `f_fnc_setAIPathing`.

Joecuronium replied:
With the locality change, it will indeed toggle.

Bubbus commented on components/zeus_ui/fn_addZeusActions.sqf:
```diff
@@ -143,9 +143,31 @@ private _actionToBeAdded =
+//Ending Dialog
+private _endingDialog =
+{
+	[]call f_fnc_endingDialog;
```
spacing ree 📏📏📏

Bubbus commented on components/zeus_ui/fn_endingDialog.sqf:
```diff
@@ -0,0 +1,33 @@
+  
+	_endingClasses pushBackUnique _class;
+	
+}forEach _endingArray;						//iterate Endings and populate dialog arrays
```
📏

Bubbus commented on components/zeus_ui/fn_endingDialog.sqf:
```diff
@@ -0,0 +1,33 @@
+	params ["_dialogValues"];
+	_dialogValues params ["_endingClass"];
+  
+  	["Executing Ending %1", _endingClass]call zen_common_fnc_showMessage;
```
📏

Bubbus commented on components/zeus_ui/fn_endingDialog.sqf:
```diff
@@ -0,0 +1,33 @@
+
+
+["Ending Dialog",
+[
```
pls indent properly 👺

Bubbus commented on components/zeus_ui/fn_endingDialog.sqf:
```diff
@@ -0,0 +1,33 @@
+ 	 [_endingClass] remoteExecCall ['f_fnc_broadcastEnding', 2];
+},
+{},
+[]]call zen_dialog_fnc_create;
```
📏📏📏

Bubbus commented on configuration/endings.hpp:
```diff
@@ -0,0 +1,75 @@
+	subtitle = "";
+	description = "Sorry!";
+	music = "FourStringEnding";				//Music will play
+	code = systemChat "Mission Broke";		//Code will run
```
This should probably be a string so that users know they need to give a string for any statements more complex than this.  For example, `code = {_x setDamage 1;} foreach allUnits; systemChat "pogging uncontrollably";` fails to compile.

```
code = "systemChat 'Mission Broke'";	//Code will run - needs to be a string.
```
```suggestion
	code = "systemChat 'Mission Broke'";    // Code will run - should be a string.
```

Joecuronium replied:
Oh no, I forgor to fix that in the PR...It was already a string at some point

Bubbus reviewed (dismissed):
Quite good changes.  I'm here to nitpick it's what I do 🤓

Joecuronium committed:
Update fn_addZeusActions.sqf

Joecuronium committed:
Update fn_endingDialog.sqf

Joecuronium committed:
Update configuration/endings.hpp

Co-authored-by: Bubbus <splambob@googlemail.com>

Joecuronium committed:
Update fn_toggleAIPath.sqf

Moved checking of current state onto _unit locality. 
State messages are now returned to the originally calling machine.

Joecuronium committed:
Update fn_zen_toggleAIPath.sqf

removed state checks

dgibso29 reviewed (dismissed):
Looks good. No further comments beyond Bubbus' review.

Joecuronium committed:
Merge branch 'develop' into feature/additional-zeus-interface-actions
